Added rule validator to verify correctness of VARSQL query

spencep1 opened the pull request:
## Changes Made
- Added `RuleGenerator.parse_validate(pattern, rewrite)` to verify query pairs as valid varsql
  - returns a tuple of [bool, str, int] that represents , bool: valid query, str: error message, int: index of error
- Added new test cases to RuleGenerator_Test, test_parse_validator_1 to test_parse_validator_13

## Testing
- All existing tests and all new tests have passed

HazelYuAhiru reviewed (commented):
Overall, this PR looks great! Just a few small comments:
- Make sure to clean up any unnecessary commented-out code, or add a brief explanation above those lines if you plan to keep them.
- Double-check for typos and make sure the newly added comments follow the same style as the rest of the codebase.

Also just to confirm, this PR only checks for rules that contain standard `SELECT`, `FROM`, and `WHERE` clauses with conditions, correct?

spencep1 commented:
## Changes Made
- Added `RuleParser.find_malformed_brackets(pattern)` to find variables that have mismatched brackets in the varSQL query
  - returns an int that represents the index of the first variable with mismatching parenthesis. It will return -1 if there are none with mismatching brackets
- Added new test cases to Rule, test_parse_validator_1 to test_parse_validator_7, which test cases with mismatching brackets

## Testing
- All existing tests and all new tests have passed

HazelYuAhiru reviewed (commented):
LGTM!
